vpr/src/pack/prepack.h

```diff
@@ -6,16 +6,175 @@
+     */
+    inline t_pb_graph_node* get_expected_lowest_cost_pb_gnode(AtomBlockId blk_id) const {
+        auto iter = expected_lowest_cost_pb_gnode.find(blk_id);
+        VTR_ASSERT(iter != expected_lowest_cost_pb_gnode.end());
+        t_pb_graph_node* pb_gnode = iter->second;
+        VTR_ASSERT(pb_gnode != nullptr);
+        return pb_gnode;
+    }
+
+    /**
+     * @brief Returns the total number of molecules in the prepacker.
+     */
+    inline size_t get_num_molecules() const {
+        size_t num_molecules = 0;
+        t_pack_molecule* molecule_head = list_of_pack_molecules;
+        for (auto cur_molecule = molecule_head; cur_molecule != nullptr; cur_molecule = cur_molecule->next) {
+            ++num_molecules;
+        }
+        return num_molecules;
+    }
+
+    /**
+     * @brief Returns all of the molecules as a vector.
+     */
+    inline std::vector<t_pack_molecule*> get_molecules_vector() const {
+        std::vector<t_pack_molecule*> molecules;
+        t_pack_molecule* molecule_head = list_of_pack_molecules;
+        for (auto cur_molecule = molecule_head; cur_molecule != nullptr; cur_molecule = cur_molecule->next) {
+            molecules.push_back(cur_molecule);
+        }
+        return molecules;
+    }
+
+    /**
+     * @brief Marks all of the molecules as valid.
+     */
+    inline void mark_all_molecules_valid() {
+        t_pack_molecule* molecule_head = list_of_pack_molecules;
+        for (auto cur_molecule = molecule_head; cur_molecule != nullptr; cur_molecule = cur_molecule->next) {
+            cur_molecule->valid = true;
+        }
+    }
+
+    /**
+     * @brief Calculates maximum molecule statistics accross all molecules,
+     */
+    t_molecule_stats calc_max_molecule_stats(const AtomNetlist& netlist) const;
+
+    /**
+     * @brief Gets the largest number of blocks that any molecule has.
+     */
+    inline size_t get_max_molecule_size() const {
+        size_t max_molecule_size = 1;
+        t_pack_molecule* molecule_head = list_of_pack_molecules;
+        for (auto cur_molecule = molecule_head; cur_molecule != nullptr; cur_molecule = cur_molecule->next) {
+            max_molecule_size = std::max<size_t>(max_molecule_size, cur_molecule->num_blocks);
+        }
+        return max_molecule_size;
+    }
+
+    /**
+     * @brief Resets the prepacker object. Clearing all state.
+     *
+     * This resets the prepacker, allowing it to prepack again and also freeing
+     * any state.
+     */
+    void reset();
+
+    /// @brief Destructor of the prepacker class. Calls the reset method.
+    ~Prepacker() { reset(); }
+
+private:
+    /**
+     * @brief A linked list of all the packing molecules that are loaded in
+     *        prepacking stage.
+     *
+     * Is is useful in freeing the pack molecules at the destructor of the Atom
+     * context using free_pack_molecules.
+     *
+     * TODO: Instead of pointers, we should use some form of ID.
```

Comment:
TODO: should also probably use a vector instead of a linked list for storage (which will give you your ids, so maybe that's already what you mean here too).

Reply:
That was the intention; but I will make the comment more clear!